### ISSUE TITLE
How to enable Fuzz-Introspector for a project?

### ISSUE THREAD

**Issue by IvanNardi** [google/oss-fuzz#8939]
Because of the recent user survey, I discover this new (new to me...) "feature": Fuzz-Introspector.
What I need to to to enable it in a project?
Reading the documentation, it seems that it should be kind of automatic, but I don't find "ndpi" in the list in https://oss-fuzz-introspector.storage.googleapis.com/index.html
What am I missing?
Sorry if this is a silly question or if it is not the right place to ask

**Comment by Navidem** [comment]
Thanks for your reach out! Fuzz Introspector enables automatically for any supported languages. 
For the case of `ndpi` it is enabled but the build is [failing](https://oss-fuzz-build-logs.storage.googleapis.com/index.html#ndpi), the most recent log is here: 
[log-6332e0dc-33a1-4516-9610-8fbf409779f0.txt](https://github.com/google/oss-fuzz/files/9954410/log-6332e0dc-33a1-4516-9610-8fbf409779f0.txt)

Looks libpcap-dev is missing?

**Comment by IvanNardi** [comment]
Thanks you very much for your help: I' ll take a look
Any specific reason why I didn't receive any notifications about the build failure? If I remember correctly, I received them if the build with "usual" sanitizers (Address/memory/undefined) fails 

Thanks again!

**Comment by Navidem** [comment]
> Any specific reason why I didn't receive any notifications about the build failure?

We did not send introspector related build failures by default as in early phases it was common to get such failures. I think we can enable it per-project.

**Comment by DavidKorczynski** [comment]
@IvanNardi I just tried to understand why the libpcap failure happens, because it's odd it happens in only fuzz introspector builds.

It seems like `AC_CHECK_LIB` complains that libpcap is not there, even though it's there.

I'm not sure about the internals of `AC_CHECK_LIB`. I wasn't able to get an accurate conclusion yet, but, the possible things I have are for why it's failing are:
- Until recently fuzz introspector's version of clang would print out some excessive output even when introspector was not used to analyse a fuzzer. If `AC_CHECK_LIB` tries to parse the output of the compilation process then this may cause it to fail. If so, we've recently upgrade oss-fuzz with the new introspector version and this failure should go away automatically.
- fuzz introspector uses LLVM LTO, thus, the `-flto` flag is set in `CFLAGS`. I'm unsure how this affects `AC_CHECK_LIB` with libpcap, but, it's often an item that has the large impact on the build process.

**Comment by IvanNardi** [comment]
Thanks @DavidKorczynski @Navidem for your insights!

This failure is very strange indeed. Buils system doesn't find a lot of libraries:
* libpcap-dev
* libjson-c ```checking for json_object_put in -ljson-c... no```
* pthreads (!!!) ```configure: WARNING: POSIX Threads not available. Building library only.```
* libnuma ```checking for numa_available in -lnuma... no```

Everything is fine with the other sanitizers.

I'll try to test `-flto` locally.
Thanks for your help

**Comment by IvanNardi** [comment]
> * we've recently upgrade oss-fuzz with the new introspector version and this failure should go away automatically.

Is that change already available in the latest image(s)?

**Comment by DavidKorczynski** [comment]
> Is that change already available in the latest image(s)?

No, just checked for the latests runs on oss-fuzz.com and the logging statement is still there.

Will update you here when I see it's gone, then we can check if ndpi passed too!

**Comment by DavidKorczynski** [comment]
@IvanNardi It is now enable but still it does not work. I did an investigation and think LTO is causing the issue.

Following some investigations, I found that the configuration fails when LTO is enabled. So, I tried disabling LTO during the configuration and enabling it again in the generated `Makefile` before building. I also commented out building of the shared library to only focus on building the static library. Am leaving the following bits in case you want to pursue investigating this:

I moved libpcap and json-c building and install into the Dockerfile:
```
# build libpcap
RUN tar -xvzf libpcap-1.9.1.tar.gz && \
  cd libpcap-1.9.1 && \
  ./configure --disable-shared && \
  make -j$(nproc) && \
  make install

RUN cd json-c && \
  mkdir build && \
  cd build && \
  cmake -DBUILD_SHARED_LIBS=OFF .. && \
  make install
```

Then, I modified `build.sh` to:

```
#Disable code instrumentation
CFLAGS_SAVE="$CFLAGS -v"
CXXFLAGS_SAVE="$CXXFLAGS -v"

# build project
# switch flags
NEW_CFLAGS_ARR="$(echo $CFLAGS | sed -e "s/-flto//g")"
NEW_CXXFLAGS_ARR="$(echo $CXXFLAGS | sed -e "s/-flto//g")"
export CFLAGS=${NEW_CFLAGS_ARR}
export CXXFLAGS=${NEW_CXXFLAGS_ARR}

# Only build static library
cd ndpi
sed -i "s/NDPI_LIBS            = \$(NDPI_LIB_STATIC) \$(NDPI_LIB_SHARED)/NDPI_LIBS            = \$(NDPI_LIB_STATIC)/g" src/lib/Makefile.in

sh autogen.sh
./configure --enable-fuzztargets

# -flto was removed during configuration to get passed it. Enable it here again.
sed -i "s/NDPI_CFLAGS =/NDPI_CFLAGS = ${CFLAGS_SAVE} #/g" ./Makefile
sed -i "s/CFLAGS =/CFLAGS = ${CFLAGS_SAVE} #/g" ./Makefile
sed -i "s/CXXFLAGS =/CXXFLAGS = ${CXXFLAGS_SAVE} #/g" ./Makefile

make
ls fuzz/fuzz* | grep -v "\." | grep -v "with_main" | while read i; do cp $i $OUT/; done
```

The resulting bits is that all goes well until something is being linked. The above fails with:
```
clang -O1 -fno-omit-frame-pointer -gline-tables-only -DFUZZING_BUILD_MODE_UNSAFE_FOR_PRODUCTION -O0 -flto -fno-inline-functions -fuse-ld=gold -Wno-unused-command-line-argument -fsanitize=fuzzer-no-link -g -v  -fPIC -DPIC -I../src/include -W -Wall -Wno-unused-parameter -Wno-unused-function -Wno-address-of-packed-member  -O2   -pthread  ndpiReader.o libndpiReader.a ../src/lib/libndpi.a -lpcap -lm   -pthread -o ndpiReader
clang version 15.0.0 (https://github.com/llvm/llvm-project.git bf7f8d6fa6f460bf0a16ffec319cd71592216bf4)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /usr/local/bin
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/9
Selected GCC installation: /usr/lib/gcc/x86_64-linux-gnu/9
Candidate multilib: .;@m64
Candidate multilib: 32;@m32
Candidate multilib: x32;@mx32
Selected multilib: .;@m64
 "/usr/bin/ld.gold" -z relro --hash-style=gnu --eh-frame-hdr -m elf_x86_64 -dynamic-linker /lib64/ld-linux-x86-64.so.2 -o ndpiReader /lib/x86_64-linux-gnu/crt1.o /lib/x86_64-linux-gnu/crti.o /usr/lib/gcc/x86_64-linux-gnu/9/crtbegin.o -L/usr/lib/gcc/x86_64-linux-gnu/9 -L/usr/lib/gcc/x86_64-linux-gnu/9/../../../../lib64 -L/lib/x86_64-linux-gnu -L/lib/../lib64 -L/usr/lib/x86_64-linux-gnu -L/usr/lib/../lib64 -L/usr/local/bin/../lib -L/lib -L/usr/lib -plugin /usr/local/bin/../lib/LLVMgold.so -plugin-opt=mcpu=x86-64 -plugin-opt=O2 --whole-archive /usr/local/lib/clang/15.0.0/lib/linux/libclang_rt.ubsan_standalone-x86_64.a --no-whole-archive --dynamic-list=/usr/local/lib/clang/15.0.0/lib/linux/libclang_rt.ubsan_standalone-x86_64.a.syms ndpiReader.o libndpiReader.a ../src/lib/libndpi.a -lpcap -lm --no-as-needed -lpthread -lrt -lm -ldl -lgcc --as-needed -lgcc_s --no-as-needed -lpthread -lc -lgcc --as-needed -lgcc_s --no-as-needed /usr/lib/gcc/x86_64-linux-gnu/9/crtend.o /lib/x86_64-linux-gnu/crtn.o
We are now in the FuzzIntrospector module pass
[Log level 1] : 11:08:06 : Fuzz introspector is running
[Log level 1] : 11:08:06 : Running introspector on ld-temp.o
[Log level 1] : 11:08:06 : Main function filename: /src/ndpi/example/ndpiReader.c
[Log level 1] : 11:08:06 : Main function is non-empty
[Log level 1] : 11:08:06 : This means a main function is in the source code rather in the libfuzzer library, and thus we do not care about it. We only want to study the actual fuzzers. Exiting this run.
clang-15: error: unable to execute command: Segmentation fault (core dumped)
clang-15: error: linker command failed due to signal (use -v to see invocation)
make[1]: *** [Makefile:54: ndpiReader] Error 254
```

Notice, the logging from fuzz-introspector indicates that fuzz-introspector did not run any of it's analysis code, which means the error happens in a pure LTO/Gold linker set up.

To summarise, in order to make fuzz-introspector work we need `ndpi` to be able to compile with LTO and LLVM gold linker.

**Comment by IvanNardi** [comment]
@DavidKorczynski , thank you very much for all your details!
I'll take a look into the issue.
Thanks again!

**Comment by IvanNardi** [comment]
Fixed in 3baf95b6ec9425509b1ea370621c82ad180a2f7a
Thanks for your help!
